File: projects/test_models/Zeta/cs/Types/ExternalModelTest.cs
using System;

namespace Zeta
{
    /// <summary>
    /// Test type
    /// </summary>
    public sealed class ExternalModelTest
    {
        #region Properties
        /// <summary>
        /// Some comment
        /// </summary>
        public Phi.Date DateTest { get; set; }
        /// <summary>
        /// Some comment
        /// </summary>
        public System.Collections.Generic.List<object> BinariesTest { get; set; }
        /// <summary>
        /// Some comment
        /// </summary>
        public System.Collections.Generic.List<System.Collections.Generic.List<object> > CollectionOfCollectionTest { get; set; }
        /// <summary>
        /// Some comment
        /// </summary>
        public System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<object> > > CollectionOfCollectionOfCollectionTest { get; set; }
        #endregion

        #region Constructors
        public ExternalModelTest() { }

        public ExternalModelTest(
            Phi.Date dateTest,
            System.Collections.Generic.List<object> binariesTest,
            System.Collections.Generic.List<System.Collections.Generic.List<object> > collectionOfCollectionTest,
            System.Collections.Generic.List<System.Collections.Generic.List<System.Collections.Generic.List<object> > > collectionOfCollectionOfCollectionTest)
        {
            DateTest = dateTest;
            BinariesTest = binariesTest;
            CollectionOfCollectionTest = collectionOfCollectionTest;
            CollectionOfCollectionOfCollectionTest = collectionOfCollectionOfCollectionTest;
        }
        #endregion

        #region Equality
        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;

            var value = obj as ExternalModelTest;
            if (value == null) return false;

            return
                DateTest == value.DateTest &&
                BinariesTest != null && value.BinariesTest != null &&
                BinariesTest.Equals(value.BinariesTest) &&
                CollectionOfCollectionTest != null && value.CollectionOfCollectionTest != null &&
                CollectionOfCollectionTest.Equals(value.CollectionOfCollectionTest) &&
                CollectionOfCollectionOfCollectionTest != null && value.CollectionOfCollectionOfCollectionTest != null &&
                CollectionOfCollectionOfCollectionTest.Equals(value.CollectionOfCollectionOfCollectionTest);
        }

        public static bool operator ==(ExternalModelTest lhs, ExternalModelTest rhs)
        {
            if (Object.ReferenceEquals(lhs, rhs))
                return true;

            return !Object.ReferenceEquals(null, lhs) && lhs.Equals(rhs);
        }

        public static bool operator !=(ExternalModelTest lhs, ExternalModelTest rhs)
        {
            return !(lhs == rhs);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                // Choose large primes to avoid hashing collisions
                const int HashingBase = (int) 2166136261;
                const int HashingMultiplier = 16777619;

                int hash = HashingBase;
                hash = (hash * HashingMultiplier) ^ DateTest.GetHashCode();
                hash = (hash * HashingMultiplier) ^
                    (!object.ReferenceEquals(null, BinariesTest) ? BinariesTest.GetHashCode() : 0);
                hash = (hash * HashingMultiplier) ^
                    (!object.ReferenceEquals(null, CollectionOfCollectionTest) ? CollectionOfCollectionTest.GetHashCode() : 0);
                hash = (hash * HashingMultiplier) ^
                    (!object.ReferenceEquals(null, CollectionOfCollectionOfCollectionTest) ? CollectionOfCollectionOfCollectionTest.GetHashCode() : 0);
                return hash;
            }
        }
        #endregion
    }
}

File: projects/test_models/Zeta/cpp/src/types/ExternalModelTest.cpp
#include "Zeta/types/ExternalModelTest.hpp"

namespace Zeta {

ExternalModelTest::ExternalModelTest()
    : DateTest_(static_cast<Phi::Date>(0)) { }

ExternalModelTest::ExternalModelTest(
    const Phi::Date DateTest,
    const std::vector<std::string>& BinariesTest,
    const std::vector<std::vector<std::string> >& CollectionOfCollectionTest,
    const std::vector<std::vector<std::vector<std::string> > >& CollectionOfCollectionOfCollectionTest)
    : DateTest_(DateTest),
      BinariesTest_(BinariesTest),
      CollectionOfCollectionTest_(CollectionOfCollectionTest),
      CollectionOfCollectionOfCollectionTest_(CollectionOfCollectionOfCollectionTest) { }

void ExternalModelTest::swap(ExternalModelTest& other) noexcept {
    using std::swap;
    swap(DateTest_, other.DateTest_);
    swap(BinariesTest_, other.BinariesTest_);
    swap(CollectionOfCollectionTest_, other.CollectionOfCollectionTest_);
    swap(CollectionOfCollectionOfCollectionTest_, other.CollectionOfCollectionOfCollectionTest_);
}

bool ExternalModelTest::operator==(const ExternalModelTest& rhs) const {
    return DateTest_ == rhs.DateTest_ &&
        BinariesTest_ == rhs.BinariesTest_ &&
        CollectionOfCollectionTest_ == rhs.CollectionOfCollectionTest_ &&
        CollectionOfCollectionOfCollectionTest_ == rhs.CollectionOfCollectionOfCollectionTest_;
}

ExternalModelTest& ExternalModelTest::operator=(ExternalModelTest other) {
    using std::swap;
    swap(*this, other);
    return *this;
}

Phi::Date ExternalModelTest::DateTest() const {
    return DateTest_;
}

void ExternalModelTest::DateTest(const Phi::Date v) {
    DateTest_ = v;
}

const std::vector<std::string>& ExternalModelTest::BinariesTest() const {
    return BinariesTest_;
}

std::vector<std::string>& ExternalModelTest::BinariesTest() {
    return BinariesTest_;
}

void ExternalModelTest::BinariesTest(const std::vector<std::string>& v) {
    BinariesTest_ = v;
}

void ExternalModelTest::BinariesTest(const std::vector<std::string>&& v) {
    BinariesTest_ = std::move(v);
}

const std::vector<std::vector<std::string> >& ExternalModelTest::CollectionOfCollectionTest() const {
    return CollectionOfCollectionTest_;
}

std::vector<std::vector<std::string> >& ExternalModelTest::CollectionOfCollectionTest() {
    return CollectionOfCollectionTest_;
}

void ExternalModelTest::CollectionOfCollectionTest(const std::vector<std::vector<std::string> >& v) {
    CollectionOfCollectionTest_ = v;
}

void ExternalModelTest::CollectionOfCollectionTest(const std::vector<std::vector<std::string> >&& v) {
    CollectionOfCollectionTest_ = std::move(v);
}

const std::vector<std::vector<std::vector<std::string> > >& ExternalModelTest::CollectionOfCollectionOfCollectionTest() const {
    return CollectionOfCollectionOfCollectionTest_;
}

std::vector<std::vector<std::vector<std::string> > >& ExternalModelTest::CollectionOfCollectionOfCollectionTest() {
    return CollectionOfCollectionOfCollectionTest_;
}

void ExternalModelTest::CollectionOfCollectionOfCollectionTest(const std::vector<std::vector<std::vector<std::string> > >& v) {
    CollectionOfCollectionOfCollectionTest_ = v;
}

void ExternalModelTest::CollectionOfCollectionOfCollectionTest(const std::vector<std::vector<std::vector<std::string> > >&& v) {
    CollectionOfCollectionOfCollectionTest_ = std::move(v);
}

}
